backend/pkg/kafka/consumer.go

```diff
@@ -99,100 +114,84 @@ func (s *Service) FetchMessages(ctx context.Context, progress IListMessagesProgr
 		kgo.ConsumePartitions(partitionOffsets),
 	)
 
-	// 4. Consume messages and check if filter code accepts the given records
+	// 4. Start go routine that consumes messages from Kafka and produces these records on the jobs channel so that these
+	// can be decoded by our workers.
+	go s.consumeKafkaMessages(workerCtx, client, consumeRequest, jobs)
+
+	// 5. Receive decoded messages until our request is satisfied. Once that's the case we will cancel the context
+	// that propagate to all the launched go routines.
 	messageCount := 0
 	remainingPartitionRequests := len(consumeRequest.Partitions)
+	for msg := range resultsCh {
```

Comment:
It's possible for this to hang here permanently. The `workerCtx` wraps a parent context, and the parent context could be canceled. If the parent context is canceled, the message workers will return, but nothing closes `resultsCh`, and this will hang here.

One way to solve this would be to have a WaitGroup underneath the workerCtx and use that for each started worker like so:
```go
go func() {
  defer wg.Done()
  s.startMessageWorker(workerCtx, isMessageOK, jobs, resultsCh)
}()
```
and then underneath the loop, another goroutine
```go
go func() {
  wg.Wait()
  close(resultsCh)
}()
```
This will ensure that `resultsCh` is closed, which ensures that this loop will stop and allow the function to return.